Implement authentication and blacklisting rules

JohnDamilola opened the pull request.

DrLynch commented on extension/extension/background.js:
```diff
@@ -9,7 +9,7 @@ var RAW_DEBUG = false;
 /* This variable must be manually updated to specify the server that
  * the data will be sent to.  
 */
-var WEBSOCKET_SERVER_URL = "wss://learning-observer.org/wsapi/in/" 
```
Why has this been removed?

JohnDamilola replied:
Do you mean the trailing whitespace? I often updated the server url and removed it the process when I noticed that there was a trailing whitespace. Was I wrong to have done that?

pmitros commented:
@bradley-erickson Can you do the review on our end?

pmitros commented:
@bradley-erickson Notes from conversation: 
1. Not all three states seem to be handled (deny for two days, or errors for future states)
2. We're not using keywords (const DENY='deny' or similar)
3. We're using in-memory storage, so DENY doesn't kill us long-term (should have FSM and use local storage or similar)
4. NCSU needs to be rebased to current code (a lot of this became a node package)
5. Rules patterns hardcode to NCSU.
6. Protocol is JSON. I'm a little confused by sending strings instead of JSON.

bradley-erickson commented on learning_observer/learning_observer/auth/events.py:
```diff
@@ -230,8 +234,12 @@ async def chromebook_auth(request, headers, first_event, source):
 
+    payload_for_validation = authdata.get('chrome_identity', {})
+    
+    auth_response = authenticate_payload(payload_for_validation)
```
What about other forms of auth that we use? Those should also be authenticated.

JohnDamilola replied:
For this, @DrLynch and @pmitros agreed that we are authenticating with just the `chromebook_auth`  auth method. My implementation is isolated for that method and does not interfare with other methods' normal working conditions.

bradley-erickson commented on learning_observer/learning_observer/auth/events.py:
```diff
@@ -321,6 +329,11 @@ async def authenticate(request, headers, first_event, source):
         if auth_metadata:
+            auth_response = auth_metadata.get('auth_response')
+            if auth_response and "status_code" in auth_response and auth_response.get("status_code") == 403:
+                print("Forbidden.")
```
This ought to be included in the logger (`debug_log()`) with more information about why. We could run into problems if the blacklisting/whitelisting pattern is broken. Including this in our debug logs could help quicker pinpoint the issue.

JohnDamilola replied:
I've replaced this print statement with `debug_log()` instead

bradley-erickson commented on learning_observer/learning_observer/incoming_student_event.py:
```diff
@@ -319,6 +319,7 @@ async def incoming_websocket_handler(request):
 
     INIT_PIPELINE = settings.settings.get("init_pipeline", True)
     json_msg = None
+    ws.send_str('Acknowledgment received')
```
I'm not sure the purpose of this acknowledgement. The only reason I kind of know what it does is because I recall you describing like this during a meeting. I think it was to make sure both the client and server had seen a message, but please expand on why its needed in a comment or docstring.

JohnDamilola replied:
Yes, it was for testing and I have removed it

bradley-erickson commented on learning_observer/learning_observer/incoming_student_event.py:
```diff
@@ -345,12 +347,19 @@ async def incoming_websocket_handler(request):
+        # Send the status message to the client (chrome extension)
+        await ws.send_json({"status": auth_response.get('type')})
+        debug_log(auth_response.get('msg'))
+        return ws
```
should we close the connection after they failed to authenticate? Does this keep the connection open? Detailing the workflow in some fashion would be useful for future devs to understand.

JohnDamilola replied:
Yes, we keep the connection open and I have put a reason for that in a comment (temporary forbidden access will need to reauthenticate in the future and any other need for reauthentication too)

bradley-erickson commented on extension/extension/background.js:
```diff
@@ -189,8 +190,13 @@ function websocket_logger(server) {
     return function(data) {
-        queue.push(data);
-        dequeue();
+        switch (authStatus) {
```
How are the 2 other cases handled? Currently we push items to the queue if we receive `allow` or `deny_for_2_days`. We should handle each one individually.

We probably want the default case to happen if we receive an unsupported case. There's a chance the server gets updated to include a new case/keyword, but the clients haven't updated the extension, so they are not sure how to handle it yet.

JohnDamilola replied:
I have handled `deny_for_2_days` while allow is the default along with any future unsupported cases from the server

bradley-erickson commented on extension/extension/background.js:
```diff
@@ -189,8 +190,13 @@ function websocket_logger(server) {
-        queue.push(data);
-        dequeue();
+        switch (authStatus) {
+            case "deny": // don't update/empty the queue if authStatus is "deny"
```
This should probably be a keyword instead similar to how you did this on the Python side `DENY = 'deny'`. Then instead checking for `case DENY:`.

JohnDamilola replied:
I have replaced every string constant with their corresponding keywords

JohnDamilola commented on learning_observer/learning_observer/auth/events.py:
```diff
@@ -348,6 +360,69 @@ def check_event_auth_config():
+
+# Priority order of rule types for sorting
+RULE_TYPES_BY_PRIORITIES = [DENY, DENY_FOR_TWO_DAYS]
+def authenticate_payload(payload):
```
I have abstracted out the code for blacklisting to a separate file and added docstrings for the files and the individual functions too

JohnDamilola commented on learning_observer/learning_observer/auth/events.py:
```diff
@@ -348,6 +360,69 @@ def check_event_auth_config():
+    DENY: [
+        {
+            "field": "email",
+            "patterns": ["^.*@ncsu.edu"]
```
I now have a `blacklisting_patterns.yaml` file for this purpose that can be setup on the different servers as we please

JohnDamilola commented on learning_observer/learning_observer/auth/events.py:
```diff
@@ -201,7 +206,6 @@ async def local_storage_auth(request, headers, first_event, source):
         'providence': 'ls'  # local storage
     }
 
-
```
Noted. I used ran the code through the linter and fixed every warning

JohnDamilola commented on learning_observer/learning_observer/auth/events.py:
```diff
@@ -331,7 +344,6 @@ async def authenticate(request, headers, first_event, source):
     print("All authentication methods failed. Unauthorized.")
     raise aiohttp.web.HTTPUnauthorized()
 
-
```
Fixed too.

JohnDamilola commented:
> @bradley-erickson Notes from conversation:
> 
> 1. Not all three states seem to be handled (deny for two days, or errors for future states)
> 2. We're not using keywords (const DENY='deny' or similar)
> 3. We're using in-memory storage, so DENY doesn't kill us long-term (should have FSM and use local storage or similar)
> 4. NCSU needs to be rebased to current code (a lot of this became a node package)
> 5. Rules patterns hardcode to NCSU.
> 6. Protocol is JSON. I'm a little confused by sending strings instead of JSON.

1. I'm handling all states now and ALLOW is the default
2. I'm now using keywords everywhere (on the client side too)
3. I refactored the code to use local storage 
4. TODO: I will rebase once the code changes has been approved
5. I've created a `blacklisting_patterns.yaml` file to handle that per server and not being hardcoded
6. I've removed the part where strings was sent from the server.
